File: scripts/push_to_github.py
#!/usr/bin/env python3
"""Push current workspace to GitHub via API (no git commit needed)."""
import os, json, base64, subprocess, urllib.request, urllib.error

TOKEN = os.environ["GITHUB_PERSONAL_ACCESS_TOKEN"].strip()
OWNER = "daviddan-241"
REPO  = "Another"
BRANCH = "main"

EXCLUDES = {
    ".git", "node_modules", "dist", ".local", ".config",
    ".pnpm-store", "__pycache__", ".cache",
}
EXCLUDE_NAMES = {"artifact.edit.toml"}
MAX_SIZE = 400_000  # 400KB

def api(method, path, body=None):
    url = f"https://api.github.com{path}"
    data = json.dumps(body).encode() if body else None
    req = urllib.request.Request(url, data=data, method=method)
    req.add_header("Authorization", f"Bearer {TOKEN}")
    req.add_header("Accept", "application/vnd.github+json")
    req.add_header("X-GitHub-Api-Version", "2022-11-28")
    req.add_header("Content-Type", "application/json")
    try:
        with urllib.request.urlopen(req) as r:
            return json.loads(r.read())
    except urllib.error.HTTPError as e:
        print(f"  ERROR {e.code} {method} {path}: {e.read().decode()[:200]}")
        return None

def collect_files(root):
    files = []
    for dirpath, dirs, filenames in os.walk(root):
        # prune excluded dirs in-place
        dirs[:] = [d for d in dirs if d not in EXCLUDES]
        for fname in filenames:
            if fname in EXCLUDE_NAMES:
                continue
            fpath = os.path.join(dirpath, fpath2 := fname)
            full = os.path.join(dirpath, fname)
            size = os.path.getsize(full)
            if size > MAX_SIZE:
                continue
            rel = os.path.relpath(full, root).replace("\\", "/")
            files.append((rel, full))
    files.sort()
    return files

def create_blob(content_bytes):
    b64 = base64.b64encode(content_bytes).decode()
    result = api("POST", f"/repos/{OWNER}/{REPO}/git/blobs", {
        "content": b64,
        "encoding": "base64"
    })
    return result["sha"] if result else None

def main():
    root = "/home/runner/workspace"

    # 1. Get HEAD
    ref = api("GET", f"/repos/{OWNER}/{REPO}/git/refs/heads/{BRANCH}")
    if not ref:
        print("Failed to get HEAD ref"); return
    parent_sha = ref["object"]["sha"]
    parent_commit = api("GET", f"/repos/{OWNER}/{REPO}/git/commits/{parent_sha}")
    base_tree_sha = parent_commit["tree"]["sha"]
    print(f"Parent: {parent_sha[:8]}  Tree: {base_tree_sha[:8]}")

    # 2. Collect files
    files = collect_files(root)
    print(f"Found {len(files)} files to push")

    # 3. Create blobs
    tree_items = []
    for i, (rel, full) in enumerate(files):
        try:
            with open(full, "rb") as f:
                content = f.read()
        except Exception as e:
            print(f"  skip {rel}: {e}")
            continue
        sha = create_blob(content)
        if sha:
            tree_items.append({"path": rel, "mode": "100644", "type": "blob", "sha": sha})
            if i % 20 == 0:
                print(f"  {i}/{len(files)} blobs created...")

    print(f"Created {len(tree_items)} blobs")

    # 4. Create tree
    tree = api("POST", f"/repos/{OWNER}/{REPO}/git/trees", {
        "base_tree": base_tree_sha,
        "tree": tree_items
    })
    if not tree:
        print("Failed to create tree"); return
    new_tree_sha = tree["sha"]
    print(f"New tree: {new_tree_sha[:8]}")

    # 5. Create commit
    commit = api("POST", f"/repos/{OWNER}/{REPO}/git/commits", {
        "message": "feat: real chat panel with Solana signing, lock/unlock, live replies + Telegram fix + Render config",
        "tree": new_tree_sha,
        "parents": [parent_sha]
    })
    if not commit:
        print("Failed to create commit"); return
    new_commit_sha = commit["sha"]
    print(f"New commit: {new_commit_sha[:8]}")

    # 6. Update ref
    result = api("PATCH", f"/repos/{OWNER}/{REPO}/git/refs/heads/{BRANCH}", {
        "sha": new_commit_sha,
        "force": False
    })
    if result:
        print(f"SUCCESS! Branch {BRANCH} updated to {new_commit_sha[:8]}")
    else:
        print("Failed to update ref")

if __name__ == "__main__":
    main()

File: scripts/push_to_github.mjs
#!/usr/bin/env node
import fs from "fs";
import path from "path";
import crypto from "crypto";

const TOKEN = process.env.GITHUB_PERSONAL_ACCESS_TOKEN?.trim();
const OWNER = "daviddan-241";
const REPO  = "Another";
const BRANCH = "main";
const MAX_SIZE = 400_000;
const CONCURRENCY = 4;

const EXCLUDE_DIRS  = new Set([".git","node_modules","dist",".local",".config",".pnpm-store","__pycache__",".cache","coverage"]);
const EXCLUDE_FILES = new Set(["artifact.edit.toml"]);

async function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

async function api(method, path2, body, retries = 3) {
  for (let attempt = 0; attempt < retries; attempt++) {
    const res = await fetch(`https://api.github.com${path2}`, {
      method,
      headers: {
        Authorization: `Bearer ${TOKEN}`,
        Accept: "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
        "Content-Type": "application/json",
      },
      body: body ? JSON.stringify(body) : undefined,
      signal: AbortSignal.timeout(30_000),
    });
    const text = await res.text();
    if (res.status === 403 && text.includes("secondary rate limit")) {
      const wait = (attempt + 1) * 8000;
      console.log(`  Rate limited — waiting ${wait/1000}s...`);
      await sleep(wait);
      continue;
    }
    if (!res.ok) { console.error(`  ERR ${res.status} ${method} ${path2}: ${text.slice(0,200)}`); return null; }
    return JSON.parse(text);
  }
  return null;
}

// Compute the git blob SHA the same way git does: "blob <size>\0<content>"
function gitBlobSha(content) {
  const header = Buffer.from(`blob ${content.length}\0`);
  const h = crypto.createHash("sha1");
  h.update(header);
  h.update(content);
  return h.digest("hex");
}

function collectFiles(root) {
  const results = [];
  function walk(dir) {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
      if (entry.isDirectory()) {
        if (!EXCLUDE_DIRS.has(entry.name)) walk(path.join(dir, entry.name));
      } else {
        if (EXCLUDE_FILES.has(entry.name)) continue;
        const full = path.join(dir, entry.name);
        const stat = fs.statSync(full);
        if (stat.size > MAX_SIZE) continue;
        results.push({ rel: path.relative(root, full).replace(/\\/g, "/"), full });
      }
    }
  }
  walk(root);
  return results.sort((a,b) => a.rel.localeCompare(b.rel));
}

// Walk the full tree in GitHub to get all existing path→sha mappings
async function fetchFullTree(treeSha) {
  const data = await api("GET", `/repos/${OWNER}/${REPO}/git/trees/${treeSha}?recursive=1`);
  if (!data) return new Map();
  const map = new Map();
  for (const item of data.tree ?? []) {
    if (item.type === "blob") map.set(item.path, item.sha);
  }
  return map;
}

async function createBlob(content) {
  const r = await api("POST", `/repos/${OWNER}/${REPO}/git/blobs`, {
    content: content.toString("base64"),
    encoding: "base64",
  });
  return r?.sha ?? null;
}

async function processInBatches(items, batchSize, fn) {
  const results = new Array(items.length);
  for (let i = 0; i < items.length; i += batchSize) {
    const batch = items.slice(i, i + batchSize);
    const batchResults = await Promise.all(batch.map((item, j) => fn(item, i + j)));
    for (let j = 0; j < batchResults.length; j++) results[i + j] = batchResults[j];
    await sleep(500); // gentle pacing between batches
  }
  return results;
}

async function main() {
  const root = "/home/runner/workspace";

  const ref = await api("GET", `/repos/${OWNER}/${REPO}/git/refs/heads/${BRANCH}`);
  if (!ref) { console.error("Cannot get HEAD ref"); process.exit(1); }
  const parentSha = ref.object.sha;
  const parentCommit = await api("GET", `/repos/${OWNER}/${REPO}/git/commits/${parentSha}`);
  const baseTreeSha = parentCommit.tree.sha;
  console.log(`Parent: ${parentSha.slice(0,8)}  Tree: ${baseTreeSha.slice(0,8)}`);

  console.log("Fetching existing tree from GitHub...");
  const existingTree = await fetchFullTree(baseTreeSha);
  console.log(`  ${existingTree.size} files already in repo`);

  const files = collectFiles(root);
  console.log(`${files.length} local files found`);

  // Only upload files whose content hash differs from what's in GitHub
  const changed = [];
  for (const f of files) {
    const content = fs.readFileSync(f.full);
    const localSha = gitBlobSha(content);
    const remoteSha = existingTree.get(f.rel);
    if (localSha !== remoteSha) {
      changed.push({ ...f, content, localSha });
    }
  }
  console.log(`${changed.length} changed files to upload`);

  if (changed.length === 0) {
    console.log("Nothing to push — already up to date.");
    return;
  }

  const treeItems = [];
  // Reuse existing blobs for unchanged files (include them in tree via their existing SHA)
  for (const [p, sha] of existingTree) {
    if (!changed.find(f => f.rel === p) && files.find(f => f.rel === p)) {
      treeItems.push({ path: p, mode: "100644", type: "blob", sha });
    }
  }

  // Upload only changed blobs
  let done = 0;
  await processInBatches(changed, CONCURRENCY, async ({ rel, content }) => {
    const sha = await createBlob(content);
    if (sha) treeItems.push({ path: rel, mode: "100644", type: "blob", sha });
    done++;
    process.stdout.write(`\r  ${done}/${changed.length} blobs uploaded`);
    return sha;
  });
  console.log();

  console.log(`Building tree with ${treeItems.length} entries...`);
  const tree = await api("POST", `/repos/${OWNER}/${REPO}/git/trees`, { tree: treeItems });
  if (!tree) { console.error("Tree creation failed"); process.exit(1); }
  console.log(`New tree: ${tree.sha.slice(0,8)}`);

  const msg = process.argv[2] || "chore: sync from Replit";
  const commit = await api("POST", `/repos/${OWNER}/${REPO}/git/commits`, {
    message: msg,
    tree: tree.sha,
    parents: [parentSha],
  });
  if (!commit) { console.error("Commit creation failed"); process.exit(1); }
  console.log(`New commit: ${commit.sha.slice(0,8)}`);

  const updated = await api("PATCH", `/repos/${OWNER}/${REPO}/git/refs/heads/${BRANCH}`, {
    sha: commit.sha,
    force: false,
  });
  if (updated) {
    console.log(`\n✅ SUCCESS — ${BRANCH} → ${commit.sha.slice(0,8)}`);
    console.log(`   https://github.com/${OWNER}/${REPO}/commit/${commit.sha}`);
  } else {
    console.error("Failed to update branch ref");
    process.exit(1);
  }
}

main().catch(e => { console.error(e); process.exit(1); });
